robolectric/src/main/java/org/robolectric/internal/bytecode/ShadowWrangler.java

```diff
@@ -44,12 +44,13 @@ public Object run(Object instance, Object roboData, Object[] params) throws Exce
   private static final MethodHandle NO_SHADOW_HANDLE = constant(Object.class, NO_SHADOW);
   private final ShadowMap shadowMap;
   private final Map<Class, MetaShadow> metaShadowMap = new HashMap<>();
-  private final Map<String, Plan> planCache = new LinkedHashMap<String, Plan>() {
-    @Override
-    protected boolean removeEldestEntry(Map.Entry<String, Plan> eldest) {
-      return size() > 500;
-    }
-  };
+  private final Map<String, Plan> planCache =
+      Collections.synchronizedMap(new LinkedHashMap<String, Plan>() {
```

Comment:
The patch is meant to introduce minimal contention if using LinkedHashMap. An LRU cache that's built for concurrency is likely to be more performant, but it may not make a significant difference overall.

I can benchmark in our codebase, since we are running ~7000 unit tests. The race condition is so bad, by the way, that almost every one of our test runs hits it.